lib/routing/sfRoute.class.php

```diff
@@ -858,4 +858,35 @@ public function unserialize($data)
     list($this->tokens, $this->defaultOptions, $this->options, $this->pattern, $this->staticPrefix, $this->regex, $this->variables, $this->defaults, $this->requirements, $this->suffix, $this->customToken) = unserialize($data);
     $this->compiled = true;
   }
+
+    public function jsonSerialize(): array
+    {
+        return [
+            'tokens'         => $this->tokens,
+            'defaultOptions' => $this->defaultOptions,
+            'options'        => $this->options,
+            'pattern'        => $this->pattern,
+            'staticPrefix'   => $this->staticPrefix,
+            'regex'          => $this->regex,
+            'variables'      => $this->variables,
+            'defaults'       => $this->defaults,
+            'requirements'   => $this->requirements,
+            'suffix'         => $this->suffix,
+            'customToken'    => $this->customToken,
+        ];
+    }
```

Comment:
Thanks to the `JsonSerializable` behaviour, this method will be automatically called any time the object is shoved at `json_encode()`. The purpose of the method is to reduce the object down to a dictionary of `key => value` pairs which best represent the current "state" of the object, returned as an associative `array`. This then becomes a trivial value to parse through `json_encode()`, and what you should be left with is a JSON-serialised array containing whatever parts of the object you want to retain for later on. Cool 👌🏻 

You'll note we are encoding _every_ property of the object into JSON, but eagle-eyed readers may also note in the following `jsonUnserialize()` method that only a subset of the cached properties are actually used when deserialisating the array back into an object. 

My reasoning for including all the properties in the cache even though we aren't using them is twofold:
- up in the `serialize()` method on lines 848-854, you can see that all the same properties are included when an object of this class gets pushed through `serialize()`, so I wanted to retain the same behaviour here for consistency; and
- I'm not super confident we have every edge case for this functionality covered. If we realise later on that we need to change how we deserialise the cached content back into an object (using more/different properties), then we already have all the known state cached and ready to use. It would be absolutely horrendous if we had to migrate from one partial set of cached state to another. 🤢